target_postgres/sinks.py

```diff
@@ -252,3 +255,62 @@ def schema_name(self) -> Optional[str]:
 
         # Schema name not detected.
         return None
+
+    def activate_version(self, new_version: int) -> None:
+        """Bump the active version of the target table.
+
+        Args:
+            new_version: The version number to activate.
+        """
+        # There's nothing to do if the table doesn't exist yet
+        # (which it won't the first time the stream is processed)
+        if not self.connector.table_exists(self.full_table_name):
+            return
+
+        deleted_at = now()
+        datetime_type = self.connector.to_sql_type(
+            {"type": "string", "format": "date-time"}
+        )  # Different from SingerSDK as we need to handle types the same as SCHEMA messsages
+        integer_type = self.connector.to_sql_type(
+            {"type": "integer"}
+        )  # Different from SingerSDK as we need to handle types the same as SCHEMA messsages
+
+        if not self.connector.column_exists(
+            full_table_name=self.full_table_name,
+            column_name=self.version_column_name,
+        ):
+            self.connector.prepare_column(
+                self.full_table_name,
+                self.version_column_name,
+                sql_type=integer_type,  # Different from SingerSDK as we need to handle types the same as SCHEMA messsages
+            )
+
+        self.logger.info("Hard delete: %s", self.config.get("hard_delete"))
+        if self.config["hard_delete"] is True:
+            self.connection.execute(
+                f"DELETE FROM {self.full_table_name} "
```

Comment:
If you look at `activate_version_hard.singer` There is an `ACTIVATE_VERSION` message before the records, and an `ACTIVATE_VERSION` message after the records. I believe right now when we get the first `ACTIVATE_VERSION` message we will end up dropping the table. 

While I think two `ACTIVATE_VERSION` messages is overkill we should support the `transferwise` ACTIVATE_VERSION style without being destructive. Need to test this

Reply:
Have a failing test added now for this. We need to do something here, maybe on the first ActivateVersion message we don't delete?

Reply:
Way back I remember seeing intermediate tables for specific versions, e.g. `users_1674486431563` so maybe the first message could be a signal to create such a temp table?

Reply:
@edgarrmondragon hmm I think the first message in the previous example is actually on purpose by the tap-postgres transfer wise implementation. I have no idea why they'd want to do that but I think they want to truncate the table before they push any data.

Ie my concern here isn't valid, I think anytime an activate version message is sent at should truncate the table (based on version)

Reply:
As I thought about this more.

We do know if a tap is outputting an `ACTIVATE_VERSION` message before sending any records. We could just ignore the first one or throw an error/warning when this happens. In my opinion this is just wrong but maybe we should treat it well as that's what `pipelinewise` does in their taps. If we started adding functionality around this behavior then I think we should expand the spec for `ACTIVATE_VERSION`

Reply:
> There is an ACTIVATE_VERSION message before the records, and an ACTIVATE_VERSION message after the records.

TL;DR - I don't think we need to do anything special in targets to handle an initial and then final `ACTIVATE_VERSION` message. We can handle both as we would either - which is to delete (or flag for deletion) any records which the version makes obsolete.

My understanding is that the initial `ACTIVATE_VERSION` message _only_ is emitted if there is no prior `STATE` bookmark for this particular stream, meaning we think we are (probably) syncing the stream for the first time. The rationale that I heard for sending twice is that there's a _potential_ optimization of making records immediately available as they land in the destination. I think the optimization benefit is dubious and I don't know if we would carry over the same in SDK-based taps. Either way, I think that's safely outside the scope of things we need to solve in this PR. 😄 

Here's a sample tap implementation where you can see the initial `ACTIVATE_VERSION` message is conditional based on the `if first_run:` check.

https://github.com/transferwise/pipelinewise-tap-postgres/blob/a97d7d87493ad5b60b7d016fecf5100f91a5c4ec/tap_postgres/sync_strategies/full_table.py#L40-L41